src/Store.spec.lua

```diff
@@ -368,39 +445,79 @@ return function()
+		it("should report an error if the listeners error when flushing", function()
+			local caughtPrevState, caughtState, caughtActionLog, caughtErrorResult
 			local mockErrorReporter = {
-				reportErrorImmediately = function(_self, message, error_)
-					reportedErrorMessage = message
-					reportedErrorError = error_
+				reportReducerError = function()
+					-- no op
 				end,
-				reportErrorDeferred = function(_self, message, error_)
-					reportedErrorMessage = message
-					reportedErrorError = error_
+				reportUpdateError = function(prevState, state, actionLog, errorResult)
+					caughtPrevState = prevState
+					caughtState = state
+					caughtActionLog = actionLog
+					caughtErrorResult = errorResult
 				end
 			}
 
+			local reducer = function(state, action)
+				if action.type == "Increment" then
+					return {
+						Value = state.Value + action.amount
+					}
+				end
+				return state
+			end
+			local store = Store.new(reducer, {
+				Value = 1,
+			}, nil, mockErrorReporter)
+
 			local innerErrorMessage = "Z4PH0D"
-			local reducerCallCount = 0
-			local reducerThatErrors = function(state, action)
-				if reducerCallCount > 0 then
+			store.changed:connect(function(state, prevState)
+				if state.Value == 15 then
 					error(innerErrorMessage)
 				end
-				reducerCallCount = reducerCallCount + 1
-			end
-			local store = Store.new(reducerThatErrors, nil, nil, mockErrorReporter)
-			expect(reportedErrorMessage).to.equal(nil)
+			end)
 
-			store:dispatch({type = "any"})
+			local actions = {
+				{type = "Increment", amount = 1},
+				{type = "Increment", amount = 3},
+				{type = "Increment", amount = 10},
+			}
+			for _, action in ipairs(actions) do
+				store:dispatch(action)
+			end
+			store:flush()
 
-			local previousAction = "previous action type was: { type: \"@@INIT\" }"
-			expect(string.find(reportedErrorMessage, innerErrorMessage)).to.be.ok()
-			expect(string.find(reportedErrorMessage, previousAction)).to.be.ok()
+			expect(caughtErrorResult.message).to.equal("Caught error flushing store updates")
 			-- We want to verify that this is a stacktrace without caring too
 			-- much about the format, so we look for the stack frame associated
 			-- with this test file
-			expect(string.find(reportedErrorError, script.Name)).to.be.ok()
+			expect(string.find(caughtErrorResult.thrownValue, script.Name)).to.be.ok()
+
+			expect(caughtActionLog[1]).to.equal(actions[1])
+			expect(caughtActionLog[2]).to.equal(actions[2])
+			expect(caughtActionLog[3]).to.equal(actions[3])
+
+			-- This is before any of the actions were processed; the flush will
+			-- apply the new state from all three actions
+			expect(caughtPrevState.Value).to.equal(1)
+			expect(caughtState.Value).to.equal(15)
+
+			store:destruct()
+		end)
+
+		it("should throw errors if dispatching while a dispatch is already happening", function()
+			local store
+			store = Store.new(function(state, action)
+				if action.type ~= "@@INIT" then
```

Comment:
Little detail but I think it'd be a little clearer if the action type matched the dispatched one
```suggestion
				if action.type == "SomeAction" then
```